### PR TITLE
don't show colormap for rgb images

### PR DESCRIPTION
**Pull request by ekdnam**
# Description

Solves issue: https://github.com/napari/napari/issues/1542

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by ekdnam** [comment]
@sofroniewn Please take a look

**Comment by sofroniewn** [comment]
Hi @ekdnam I just gave this a try and i could reproduce the same problem you were mentioning about the colorbar appearing in the top left, which is likely due to the parent not being set properly. See

Screen Shot 2020-08-27 at 9 25 43 PM

Let's try a different approach. Instead of not showing the widgets, let's just hide them. I think that will be something like

```python
        if layer.rgb:
            self.grid_layout.itemAtPosition(3, 0).widget().hide()
            self.colorbarLabel.hide()
            self.colormapComboBox.hide()
```

Where we need to hide the label, the colorbar, and the combobox.

For ease of reading do you want to try putting this at the end of the init method (after everything gets added). You can also remove the extra commented lines you added above.

I think this should work. Let me know if you run into more problems!!

**Comment by ekdnam** [comment]
@sofroniewn I have made the changes, please take a look. 

the line `self.grid_layout.itemAtPosition(3, 0).widget().hide()` led to an error, so I removed it, and the code works just fine.

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 1]
```diff
@@ -98,8 +98,12 @@ def __init__(self, layer):
         self.grid_layout.addWidget(self.contrastLimitsSlider, 1, 1)
         self.grid_layout.addWidget(QLabel('gamma:'), 2, 0)
         self.grid_layout.addWidget(self.gammaSlider, 2, 1)
-        self.grid_layout.addWidget(QLabel('colormap:'), 3, 0)
-        self.grid_layout.addLayout(colormap_layout, 3, 1)
+
+        # The colormap widget won't be created when image has a rgb layer
+        if layer.rgb is False:
+            self.grid_layout.addWidget(QLabel('colormap:'), 3, 0)
+            self.grid_layout.addLayout(colormap_layout, 3, 1)
```
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_image_controls.py]: I'd undo these changes, I think if we are going with the hide approach it is fine to add them and then hide them.

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 2]
```diff
@@ -98,8 +98,12 @@ def __init__(self, layer):
         self.grid_layout.addWidget(self.contrastLimitsSlider, 1, 1)
         self.grid_layout.addWidget(QLabel('gamma:'), 2, 0)
         self.grid_layout.addWidget(self.gammaSlider, 2, 1)
-        self.grid_layout.addWidget(QLabel('colormap:'), 3, 0)
-        self.grid_layout.addLayout(colormap_layout, 3, 1)
+
+        # The colormap widget won't be created when image has a rgb layer
+        if layer.rgb is False:
+            self.grid_layout.addWidget(QLabel('colormap:'), 3, 0)
```
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_image_controls.py]: let's actually store the reference to the colormap text label explicitly by making `self.colormapLabel = QLabel('colormap:')`. Can you do that here https://github.com/napari/napari/blob/master/napari/_qt/layer_controls/qt_image_controls_base.py#L56

And then here can call `self.grid_layout.addWidget(self.colormapLabel, 3, 0)`

You can also then change this line accordingly too https://github.com/napari/napari/blob/master/napari/_qt/layer_controls/qt_surface_controls.py#L39

Having this reference to the QLabel will make hiding the label much cleaner

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 3]
```diff
@@ -114,6 +118,11 @@ def __init__(self, layer):
         self.grid_layout.setColumnStretch(1, 1)
         self.grid_layout.setSpacing(4)
 
+        # Similarly, the colormap combo box will be hidden if the layer is rgb
+        if layer.rgb:
+            self.colorbarLabel.hide()
```
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_image_controls.py]: If you see my other comment about leaving the adding of the widgets and creating `self.colormapLabel`, then here we just need to add `self.colormapLabel.hide()` and everything should work fine

**Review by sofroniewn — CHANGES_REQUESTED** [review]
Hi @ekdnam this is coming along very nicely I think. I've requested a couple changes to make us only use the hide functionality.

We'll also want to add a dedicated test for this functionality. Basically you'll want to create one test with a non-rgb layer and check that these widgets are visible, and then you'll want to create another test with an rgb layer and check they are hidden.

You'll want to put the two new tests in a new file called `test_qt_image_layer.py` which will be very similar to this file, but which will for now only contain those two tests. https://github.com/napari/napari/blob/master/napari/_qt/layer_controls/_tests/test_qt_image_base_layer_.py

**Comment by d-v-b** [comment]
In the interest of maintaining UX uniformity, why not keep the colormap dropdown but add "RGB" to it? Then someone can easily switch between viewing an image with shape (..., [3,4])-sized image as RGB or separate slices. When viewing an image where the dimensions don't permit a conversion to RGB, the "RGB" option is grayed out but still visible in the dropdown.

**Comment by ekdnam** [comment]
@d-v-b I tried something different, now for RGB images the widgets would be turned off, by,

```
if layer.rgb is True:
            self.colorbarLabel.setDisabled(True)
            self.colormapComboBox.setDisabled(True)
```

The functionality of the colorbar label and the combo box gets disabled (they cannot be used now), but they don't become _grey_ per se.

Screenshot attached below: 

napari-disabled

**Comment by ekdnam** [comment]
@sofroniewn @jni @tlambert03 @kne42 is disabling the widgets also an option that can be implemented here?

**Comment by sofroniewn** [comment]
> In the interest of maintaining UX uniformity, why not keep the colormap dropdown but add "RGB" to it? Then someone can easily switch between viewing an image with shape (..., [3,4])-sized image as RGB or separate slices. When viewing an image where the dimensions don't permit a conversion to RGB, the "RGB" option is grayed out but still visible in the dropdown.

@d-v-b We were planning to expose those sorts of conversions in the layer list, rather than the colormap dropdown as they will change the layer representation, but it is an interesting idea, and maybe one we can follow up on. It is unclear to me how to undo that operation cleanly from that dropdown menu.

An alternative might just be to show `RGB` as a string in the combo box and disable it entirely as was done by @ekdnam here

> The functionality of the colorbar label and the combo box gets disabled (they cannot be used now), but they don't become grey per se.

We could also add hide the colorbar icon.

The following seems like a reasonable improvement on where we are now and will be easy to implement

- hide the colorbar
- set the combobox to read RGB
- disable the combobox so it can't be changed

I think we can think about more complex functionality in the future

**Comment by ekdnam** [comment]
@sofroniewn should I move ahead with your suggestions?

Namely,

> hide the colorbar
> set the combobox to read RGB
> disable the combobox so it can't be changed

**Comment by ekdnam** [comment]
@d-v-b what do you think?

**Comment by d-v-b** [comment]
I agree with @sofroniewn

**Comment by sofroniewn** [comment]
> @sofroniewn should I move ahead with your suggestions?

Go for it @ekdnam

**Comment by ekdnam** [comment]
The ui now looks like this

combobox

**Comment by ekdnam** [comment]
Quite a few tests are failing. What should be done to pass 'em?

**Comment by tlambert03** [comment]
> Quite a few tests are failing. What should be done to pass 'em?

two issues:
- you need to give you QLabel a parent, so that it is properly cleaned up when the widget is destroyed:
`self.colormapLabel = QLabel('colormap:', parent=self)`
- surface `add_layer` tests are failing because `QtSurfaceControls` inherits `QtBaseImageControls` ... but surface layers have no attribute `rgb`, so this code fails for surface layers:
https://github.com/napari/napari/blob/0c9022ff73a449f9fe366e8cef879cd9909e2431/napari/_qt/layer_controls/qt_image_controls_base.py#L54-L58

you could check for `if hasattr(layer, 'rgb')` ... but I wonder whether this isn't indicative of a deeper design issue.  @sofroniewn ?

**Comment by sofroniewn** [comment]
> you could check for if hasattr(layer, 'rgb') ... but I wonder whether this isn't indicative of a deeper design issue. @sofroniewn ?

Agreed indicative of a large issue that we might want to resolve when do things like the `ColorManger` clean up see #1356, so I think it's probably ok to use that quick fix for now.

Thinking about this more @ekdnam I wonder if instead of adding `RGB` to the combobox and disabling it, maybe we should just hide the combobox and replace it with a `QLabel` with RGB instead, then it will be clear that would can't click on the combobox at all.

I worry now that people might try and click the combobox and be confused about why it is disabled

**Comment by sofroniewn** [comment]
Also in conversations @jni made the suggestion to make the label `(RBG)` instead as it isn't a colormap in quite the same way that the others are

**Comment by ekdnam** [comment]
Okay, I will add a QLabel for RGB images, and also `parent=self`.

**Comment by liu-ziyang** [comment]
@ekdnam Hi there, I created another PR to update your proposed change, you may need to pull master to your branch to see the clean change list, hopefully this unblocks your PR: https://github.com/ekdnam/napari/pull/1

**Comment by ekdnam** [comment]
@ziyangczi  Okay, will rebase my branch with master, and also look into your changes. Thanks!

**Comment by ekdnam** [comment]
What should be done to solve these errors? cc @sofroniewn @ziyangczi

File: napari/_qt/layer_controls/qt_surface_controls.py [hunk 4]
```diff
@@ -36,7 +36,7 @@ def __init__(self, layer):
         self.grid_layout.addWidget(self.contrastLimitsSlider, 1, 1)
         self.grid_layout.addWidget(QLabel('gamma:'), 2, 0)
         self.grid_layout.addWidget(self.gammaSlider, 2, 1)
-        self.grid_layout.addWidget(QLabel('colormap:'), 3, 0)
+        self.grid_layout.addWidget(self.colormapLabel, 3, 0)
```
Comment (liu-ziyang) [inline napari/_qt/layer_controls/qt_surface_controls.py]: I believe this is the line causing issue, there is no colormapLabel in the object, you may want to revert it

**Comment by ekdnam** [comment]
@ziyangczi thanks! all checks have been passed.
@sofroniewn can this PR be merged now?

**Review by sofroniewn — APPROVED** [review]
Yes this looks good to me. Here's what the final product looks like, very nice.

Thanks for the work here @ekdnam and thanks @ziyangczi for help finishing this up! I will merge now

Screen Shot 2020-10-20 at 8 28 52 AM

**Pull request merged**